### PR TITLE
Refine typesetting

### PR DESCRIPTION
**Pull request by cgcostume**

### CODE REVIEW

File: source/text/label.ts [hunk 1]
```diff
@@ -16,190 +16,77 @@ import { Text } from './text';
...
 
-    /**
-     * Returns, whether or not the character at a given index is equal to the default or the text's line feed character.
-     * @param index - The zero-based index of the desired character. If there is no character at the specified index,
-     * NaN is returned.
-     * @returns - true if char at given index equals the text's line feed character
-     */
-    lineFeedAt(index: number): boolean {
-        return this.charAt(index) === this.lineFeed;
+    set wrap(flag: boolean) {
```
Comment (anne-gropler) [inline source/text/label.ts]: missing doc, same for elide and ellipsis

File: source/text/label.ts [hunk 2]
```diff
@@ -16,190 +16,77 @@ import { Text } from './text';
...
 
-    /**
-     * Character that is to be used for Line feed.
-     */
-    get lineFeed(): string {
-        if (this._text instanceof Text) {
-            return this._text.lineFeed;
-        }
-        return Text.DEFAULT_LINEFEED;
+    set lineWidth(lineWidth: number) {
```
Comment (anne-gropler) [inline source/text/label.ts]: missig doc.
it is important that the setter sets the line width in Label.SpaceUnit, while the getter returns the line width in Typesetting unit.

File: source/text/label.ts [hunk 3]
```diff
@@ -320,19 +200,19 @@ export class Label {
      * screen space, world space, ...). This can be set either explicitly or implicitly using various transformation
      * utility functions.
      */
-    set transform(transform: mat4) {
-        if (mat4.equals(this._transform, transform)) {
+    set staticTransform(transform: mat4) {
```
Comment (anne-gropler) [inline source/text/label.ts]: add in doc: why is it called `static`? Can it be updated by the user/renderer?

File: source/text/label.ts [hunk 4]
```diff
@@ -392,6 +406,21 @@ export class Label {
 
 export namespace Label {
 
+    export enum Elide {
```
Comment (anne-gropler) [inline source/text/label.ts]: why is there `Elide` and `WordWrap`? Is any combination possible? (WordWrap seems unused)

File: source/text/typesetter.ts [hunk 5]
```diff
@@ -19,199 +26,273 @@ import { Label } from './label';
...
-     * The given vertical extent is increased by the label's line height and the horizontal extent is set to either
-     * 0 or the extent of the non-depictable characters.
-     * @param label - Label to compute accumulated advances for.
-     * @param index - In/out parameter: index to backtrack the advance of non-depictable glyphs.
-     * @param begin - Index to start glyph advance accumulation at.
-     * @param pen - In/out parameter: pen (typesetting position) to be used and adjusted.
-     * @param extent - In/out parameter: extent to be adjusted.
+     * Resolves a typed float array storing the advances of each of the label's characters.
+     * @param label - Label to resolve advances for.
+     * @param text - Text to compute advances for, if none is fiven, label.text is used.
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: fiven -> given. This typo is unforfivable!

File: source/text/typesetter.ts [hunk 6]
```diff
@@ -19,199 +26,273 @@ import { Label } from './label';
...
     /**
-     * Revert advance of preceding, not depictable glyphs. Intended to be used, e.g., on line feed.
-     * The given vertical extent is increased by the label's line height and the horizontal extent is set to either
-     * 0 or the extent of the non-depictable characters.
-     * @param label - Label to compute accumulated advances for.
-     * @param index - In/out parameter: index to backtrack the advance of non-depictable glyphs.
-     * @param begin - Index to start glyph advance accumulation at.
-     * @param pen - In/out parameter: pen (typesetting position) to be used and adjusted.
-     * @param extent - In/out parameter: extent to be adjusted.
+     * Resolves a typed float array storing the advances of each of the label's characters.
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: I am not sure if 'resolve' is the right word. Sounds like resolving a problem. Did you mean 'return' , 'compute', 'extract', ..?

Comment (anne-gropler) [reply source/text/typesetter.ts]: same for kernings()

File: source/text/typesetter.ts [hunk 7]
```diff
@@ -19,199 +26,273 @@ import { Label } from './label';
...
-     * Adjusts the vertices for a line after typesetting (done due to line feed, word wrap, or end of line) w.r.t.
-     * the targeted line alignment.
-     * @param pen - Current typesetting position (probably the end of the line in typesetting space).
-     * @param alignment - Targeted alignment, e.g., left, center, or right.
-     * @param glyphs - Glyph vertices for rendering to align the origins' x-components of (expected untransformed).
-     * @param begin - Vertex index to start alignment at.
-     * @param end - Vertex index to stop alignment at.
+     * Resolves a typed float array storing the kernings of each of the label's characters.
+     * @param label - Label to resolve kernings for.
+     * @param text - Text to compute advances for, if none is fiven, label.text is used.
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: fiven.
5, fiver, fiven. :P

File: source/text/typesetter.ts [hunk 8]
```diff
@@ -19,199 +26,273 @@ import { Label } from './label';
...
+    }
 
-        /* Origin is expected to be in typesetting space (not transformed yet). */
-        for (let i = begin; i <= end; ++i) {
-            glyphs.vertices[i].origin[0] += penOffset;
+
+    /**
+     * Create array of word, delimiter, and line feed fragments. A fragment thereby denotes the start and exclusive end
+     * index as well as the type. The array is intended to favor maintainability over performance.
+     * @param label -
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: We all know what a label is by now. ^^

File: source/text/typesetter.ts [hunk 9]
```diff
@@ -19,199 +26,273 @@ import { Label } from './label';
...
     /**
-     * Adjusts the vertices for line anchor (done due after typesetting) w.r.t. the targeted anchoring.
-     * @param label - Label to adjust the y-positions for.
-     * @param glyphs - Glyph vertices for rendering to align the origins' y-components of (expected untransformed).
-     * @param begin - Vertex index to start alignment at.
-     * @param end - Vertex index to stop alignment at.
-     *
-     * @todo Apply once at the beginning! Initial offset!
+     * Compute fragment widths without kernings w.r.t. preceding and subsequent fragments.
+     * @param fragments -
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: Who let the docs out?

File: source/text/typesetter.ts [hunk 10]
```diff
@@ -19,199 +26,273 @@ import { Label } from './label';
...
-            case Label.LineAnchor.Descent:
-                offset = label.fontFace!.descent;
-                break;
-            case Label.LineAnchor.Top:
-                offset = label.fontFace!.base;
-                break;
-            case Label.LineAnchor.Bottom:
-                offset = label.fontFace!.base - label.fontFace!.lineHeight;
+    /**
+     * Computes the left and right side thresholds for elide computation.
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: At some point, there should be an explanation of what 'elide' means. What do we elide, and what for?

File: source/text/typesetter.ts [hunk 11]
```diff
@@ -19,199 +26,273 @@ import { Label } from './label';
...
     }
 
 
     /**
      * Computes origin, tangent, and up vector for every vertex of in the given range.
      * @param transform - Transformation to apply to every vertex.
-     * @param glyphs - Glyph vertices to be transformed (expected untransformed, in typesetting space).
+     * @param vertices - Glyph vertices to be transformed (expected untransformed, in typesetting space).
      * @param begin - Vertex index to start alignment at.
      * @param end - Vertex index to stop alignment at.
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: doc: start/begin *alignment*? or transform?

File: source/text/typesetter.ts [hunk 12]
```diff
@@ -19,199 +26,273 @@ import { Label } from './label';
...
-        if (glyphs === undefined || mat4.equals(transform, mat4.create())) {
+    private static transformVertex(transform: mat4,
+        vertices: GlyphVertices | undefined, begin: number, end: number): void {
+        if (vertices === undefined || mat4.equals(transform, mat4.create())) {
             return;
         }
 
         for (let i: number = begin; i < end; ++i) {
-            const v = glyphs.vertices[i];
+            const v = vertices.vertices[i];
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: I understand the why you renamed 'glyphs' to 'vertices', but now this line looks funny :D

File: source/text/typesetter.ts [hunk 13]
```diff
@@ -226,84 +307,192 @@ export class Typesetter {
...
+            vertices.vertices[i].origin[0] += offset;
+        }
+    }
 
     /**
-     * Transforms the labels extent (typesetting space to, e.g., world space or screen space).
-     * @param transform - Transformation that was applied to every vertex.
-     * @param extent - Untransformed label extent (in typesetting space) to be transformed.
-     * @returns - Transformed label extent (copy).
+     * Create and transform glyph vertices for rendering.
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: doc: where does the *created* vertices go? What exactly do we *transform* here?
Furthermore, it helps to document when a param is a in/out param.

File: source/text/typesetter.ts [hunk 14]
```diff
@@ -226,84 +307,192 @@ export class Typesetter {
...
-     * @param glyphs the glyph vertices, a prepared (optionally empty) vertex storage
-     * @param begin vertex index to start the typesetting (usually 0)
+     * @param label - The label that is to be typeset.
+     * @param vertices - The glyph vertices, a prepared (optionally empty) vertex storage.
      */
-    static typeset(label: Label, glyphs?: GlyphVertices, begin?: number): GLfloat2 {
-        assert(!!label.fontFace, `expected a font face for label before typesetting`);
+    static typeset(label: Label, vertices: GlyphVertices): void {
+
+        /** @todo its  wrong to get the vertices here when we do not even know how many glyphs will be typeset. */
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: is this still *todo*?

File: source/text/typesetter.ts [hunk 15]
```diff
@@ -226,84 +307,192 @@ export class Typesetter {
...
+            return;
+        }
+
+        assert(label.fontFace !== undefined, `expected a font face for label before typesetting`);
+        const fontFace = label.fontFace!;
+
+        /* Retrieve advances, kernings, as well as line feed and delimiter indices. */
+
+        const glyphs = (index: number): Glyph => index < label.length ? fontFace.glyph(label.charCodeAt(index)) :
+            fontFace.glyph(label.ellipsis.charCodeAt(index - label.length));
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: I don't understand the semantics behind the `index - label.length` math. The `ellipsis` is supposed to be a very short character sequence, right?

File: source/text/typesetter.ts [hunk 16]
```diff
@@ -226,84 +307,192 @@ export class Typesetter {
...
-        for (; index !== iEnd; ++index) {
-            const glyph = label.fontFace!.glyph(label.charCodeAt(index));
+                /* Elide takes precedence, since full line width is used, so every line break is omitted. */
+                const lineFeed = !elide && fragment[2] === Typesetter.FragmentType.LineFeed;
+                let wordWrap = false;
 
-            /* Handle line feeds as well as word wrap for next word (or next glyph if word exceeds the line width). */
-            const feedLine = label.lineFeedAt(index) || (label.wordWrap &&
-                Typesetter.wordWrap(label, pen, glyph, index, safeForwardIndex));
+                /* If word wrap is desired (no elide, no line feed already and label enabled), then keep words with
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: keep -> take (?)

File: source/text/typesetter.ts [hunk 17]
```diff
@@ -226,84 +307,192 @@ export class Typesetter {
...
+                subsequent delimiters into account. These should be moved to the next line together (not split). */
+                if (!elide && !lineFeed && label.wrap) {
+                    /* If this fragment is a non depictable delimiter keep don't break. */
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: can you please re phrase the comment?

**Comment by anne-gropler** [comment]
As I understand it, if there is Elide, then line feed is ignored. This is ok, but might look different from what the user might expect. Example:
`A\nwonderful`
becomes
`Awonderful`
instead of
```
A
wonderful
```

I'd expect that at least it resolves to `A wonderful`.

**Comment by anne-gropler** [comment]
Change Request: if `lineWidth` is zero, then not even the Ellipsis should be rendered.

**Comment by anne-gropler** [comment]
For short Labels, `Middle` und `Left` don't seem to work.

File: source/text/label.ts (hunk 3 above)
Comment (cgcostume) [reply source/text/label.ts]: next commit will show that i already started on  separating transforms in dynamic and static transforms. A label has to be specified as dynamic or static on construction. When static it behaves as is currently implemented, i.e., all transformations are baked into the glyph vertices. Dynamic labels will allow to only bake the minimal required transforms and allows to update position, rotation, etc. without the need of re-computation or re-typesetting. This, however, will be a little bit slower while rendering.

File: source/text/label.ts [hunk 18]
```diff
@@ -52,35 +228,50 @@ export class Label {
...
     }
     get ellipsis(): string {
         return this._ellipsis;
     }
 
-
+    /**
+     * Line width used to either maximum length for elide or maximum length for line breaks due to word wrap. The line
+     * width is expected in typesetting space (the unit used while Typesetting, i.e., the unit as the font face's glyph
+     * texture atlas).
```
Comment (anne-gropler) [inline source/text/label.ts]: @cgcostume 
This is wrong documented. The lineWidth setter does *not* expect the line width in typesetting space, but in the space unit used for font size and positioning. (Have a look at the lineWidth getter implementation and documentation).

**Pull request merged**